game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java

```diff
@@ -55,13 +57,15 @@ public void setResourceLoader(
       final int initialUnitWidth,
       final int initialUnitHeight,
       final int initialUnitCounterOffsetWidth,
-      final int initialUnitCounterOffsetHeight) {
+      final int initialUnitCounterOffsetHeight,
+      final MapData mapData) {
```

Comment:
I'm still pretty concerned with the new MapData  dependency being wired into the class. It strikes me as similar to the swing 'parent' component dependency that is wired throughout the codebase and we're having a lot of trouble now because of it. 

I don't think there is any better time than now to avoid the new dependency. It also would seem a lot cleaner to have a focused strategy object take its place and avoid adding more rendering code to unit image factory. The unit highlight is in the same league, but at least it does not introduce a very heavy, new dependency.

Reply:
While I don't particularly like injecting `MapData` here, I'm not really seeing a quick win to avoid it in a way consistent with how the classes are structured at the moment. The only thing I can imagine doing without a major refactor is instead create a new class (something like `UnitImageRender`) and inject `MapData` into that then inject that class into `UnitImageFactory` instead. 

My general feel around that is it ends up being a half baked refactoring where a lot of the other methods in `UnitImageFactory` need to be included in that refactoring to essentially split what `UnitImageFactory` does now into 2 separate functions. And even doing that you still end up with having to inject `MapData` into that new class unless you also go and do a major refactoring to `MapData` to split it into smaller pieces.

Reply:
> The only thing I can imagine doing without a major refactor is instead create a new class (something like UnitImageRender) and inject MapData into that then inject that class into UnitImageFactory instead.

Yes! :smile: 


>  half baked refactoring where a lot of the other methods in UnitImageFactory need to be included in that refactoring to 

You have a good point, but the other elements in UnitImageFactory would not necessarily go into that same strategy object, but different ones. Thinking about SRP, it woudl be just as inappropriate to have one class know how to do every kind of image rendering, just as it's not "appropriate" for UnitImageFactory to know how to do all image rendering.

I think the wins with the half baked refactoring are:
- we stop the bleeding
- do not make an existing problem worse
- we would not be doing any kind of throw-away work, the extracted code would not need to be touched, the existing code would then build on that pattern and continue isolating and splitting functionality in a similar pattern

> And even doing that you still end up with having to inject MapData into that new class unless you also go and do a major refactoring to MapData to split it into smaller pieces.

True, though I think having just the one dependency on MapData will be easier to deal with compared to having many parameters and mapData being one of them.
 
>  If there is a quick win here I'm all for it but everything I see ends up a major effort or it ends up half baked.

I think the half baked version is the quick win. I view it much less as half baked compared to not making an existing problem worse.

Reply:
I guess we just see different value in the half baked version vs consistency. I feel like the minimal improvements are outweighed by the inconsistency it creates across the rest of the class. This can then cause confusion for the next developer that reads this class and wonders "why is the colorizing check logic in a separate class but the highlighting logic is in this class?" If classes are frequently getting updated and we believe the rest of the class would be refactored in short order then I see the benefit of the short term inconsistency to create the pattern to follow. But when the reality is it probably won't be touched again for a long time, we then live with the inconsistency for a long time which actually makes the code base harder to work with.

Reply:
Hmm, reading this class, my reaction is, "oh damn.. the unit highlight code is here". Having other, unrelated rendering logic extracted would leave the same reaction and avoid two WTFs of "oh wtf, all sorts of rendering code is here"

So I don't see it as half baked at all, it's strategically inserting new code without modifying a bunch of existing code. In general adding various if statements and a large chunk of new logic to existing code is a design smell (open-closed principle violation, probably SRP violation as well).

I fear that if we don't come back to this soon, then it'll be even harder to refactor later and it'll be yet more work to do. Worse yet, if we add more rendering logic, we could apply the same logic and have a third thing.. And then there is more dependency on mapData, maybe we bring in a GameData object... 

I really just don't see the inconsistency, having all rendering code in one place is not consistent IMO, it's just a complete lack of modularity and organization of code.

Reply:
Another point of view, `ImageColorizer` I kinda liked; with ImageTransformer it actually seems more inconsistent now than before. At a bit of a deeper level really. UnitImageFactory has several operations of 'image->image', and ImageTransformer has similar. So we have scaling and highlight in unit image factory, and we have coloring and flipping in ImageTransformer. 

I suppose to really be consistent it would all be one class.. Personally I have *zero* problem with small, focused classes; having small and easy to understand, well tested modules IMO is a good way to go. To that extent I kinda liked "ImageColorizer" as a class and would have pulled out the 'flip' code to something like "ImageRotator"

With the above perspective, could be suggested that have any external class at all is not consistent. I'm spotting in this review that I suspect it could be simpler to extract even more from UnitImageFactory, and the extracted and new code perhaps might not be just a single new class.

Reply:
I actually think that the other image transformation functions in `UnitImageFactory` potentially belong in `ImageTransformer` like scaling/highlighting as these are generic image transformations and IMO fit well together. I don't really think you need a separate class for each of these given that they are closely related and fairly small amount of code. Now if eventually `ImageTransformer` became too large then you could break it into smaller pieces.